### ISSUE TITLE
event_enable_debug_mode requires evthreads

### ISSUE THREAD

**Issue by dormando** [libevent/libevent#1720]
While trying to debug something (in my crazy super old app that uses things like `event_assign` for speed and `event_set` sometimes because old and etc), I kept getting nonsense assert errors after enabling `event_enable_debug_mode`

We use libevent with multiple threads, but not in thread mode. IE: Each worker thread gets a dedicated event base, and _only the owner thread_ ever manipulates that event base. I understood that this should be fine?

The debug code relies on a globally locked tree to track events, but locks aren't enabled in my case, so it will fail when unrelated event bases are being manipulated. I've enabled locks so I can run the debug mode (which seems happy now) but now I'm not sure if I've fixed my bug or if adding locks somehow changes my code's behavior.

Am I using it wrong (base per thread + no locks) or would it be reasonable to get debug mode's lock set up independently of enabling them for all event bases? Or perhaps on a per-base basis?

Thanks!

**Comment by azat** [comment]
Yes, your analysis is correct, debug mode requires lock, and it is global.

>Or perhaps on a per-base basis?

This make sense to me, though the problem is that event can have no base, i.e. in `event_assign`, though I guess we can add separate code path for it, I don't remember exactly.

>or would it be reasonable to get debug mode's lock set up independently of enabling them for all event bases? 

This will work, but I'm not sure about the code complexity for this case.
